agent/src/main/scala/za/co/absa/atum/agent/model/Measurement.scala

```diff
@@ -16,4 +16,62 @@
 
 package za.co.absa.atum.agent.model
 
-case class Measurement(measure: Measure, result: Any)
+import za.co.absa.atum.agent.exception.MeasurementProvidedException
+import za.co.absa.atum.model.dto.MeasureResultDTO.ResultValueType
+
+trait Measurement {
+  val measure: Measure
+  val resultValue: Any
+  val resultType: ResultValueType.ResultValueType
+}
+
+/**
+ * When the application/user of Atum Agent provides actual results by himself, the type is precise and we don't need
+ * to do any adjustments.
+ */
+case class MeasurementProvided[T](measure: Measure, resultValue: T, resultType: ResultValueType.ResultValueType)
+  extends Measurement
+
+object MeasurementProvided {
+
+  def handleSpecificType[T](
+   measure: Measure, resultValue: T, requiredType: ResultValueType.ResultValueType
+  ): MeasurementProvided[T] = {
+
+    val actualType = measure.resultValueType
+    if (actualType != requiredType)
+      throw MeasurementProvidedException(
+        s"Type of a given provided measurement result and type that a given measure supports are not compatible! " +
+          s"Got $actualType but should be $requiredType"
+      )
+    MeasurementProvided[T](measure, resultValue, requiredType)
+  }
+
+  def apply[T](measure: Measure, resultValue: T): Measurement = {
+    resultValue match {
+      case l: Long =>
+        handleSpecificType[Long](measure, l, ResultValueType.Long)
+      case d: Double =>
+        handleSpecificType[Double](measure, d, ResultValueType.Double)
+      case bd: BigDecimal =>
+        handleSpecificType[BigDecimal](measure, bd, ResultValueType.BigDecimal)
+      case s: String =>
+        handleSpecificType[String](measure, s, ResultValueType.String)
+
+      case unsupportedType =>
+        val className = unsupportedType.getClass.getSimpleName
+        throw MeasurementProvidedException(
+          s"Unsupported type of measurement for measure ${measure.measureName}: $className " +
+            s"for provided result: $resultValue"
+        )
+    }
+  }
+}
+
+/**
+ * When the Atum Agent itself performs the measurements, using Spark, then in some cases some adjustments are
+ * needed - thus we are converting the results to strings always - but we need to keep the information about
+ * the actual type as well.
+ */
+case class MeasurementByAtum(measure: Measure, resultValue: String, resultType: ResultValueType.ResultValueType)
+  extends Measurement
```

Comment:
Do we need these two classes?
If we do, please put them into a companion object to keep the in this file.

Reply:
Yes they are needed; why - what benefit would it have? I think that in this case it would only slightly complicate their usages - if the only reason is encapsulation, then they are, after all, in their own module 🤔

Reply:
I am old school 👼 , don't like top level classes present in a file named by another class.
Also they are `Measurement` descendants, so I see no problem to hide them inside. 
Usage is not an issue with import.